addons/product/models/product_template.py

```diff
@@ -938,6 +939,83 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
         """
         return next(self._get_possible_combinations(parent_combination, necessary_values), self.env['product.template.attribute.value'])
 
+    def _cartesian_product(self, product_template_attribute_values, parent_combination):
+        """
+        Generate all possible combination for attributes values (aka cartesian product).
+        It is equivalent to itertools.product except it skips invalid partial combinations before they are complete.
+
+        Imagine the cartesian product of 'A', 'CD' and range(1_000_000) and let's say that 'A' and 'C' are incompatible.
+        If you use itertools.product or any normal cartesian product, you'll need to filter out of the final result
+        the 1_000_000 combinations that start with 'A' and 'C' . Instead, This implementation will test if 'A' and 'C' are
+        compatible before even considering range(1_000_000), skip it and and continue with combinations that start
+        with 'A' and 'D'.
+
+        It's necessary for performance reason because filtering out invalid combinations from standard Cartesian product
+        can be extremely slow
+
+        :param product_template_attribute_values: the values we want all the possibles combinations of.
+        One list of values by attribute line
+        :return: a generator of product template values combinations (in the form of lists)
+        """
+        if not product_template_attribute_values:
+            return
+
+        all_exclusions = self._get_own_attribute_exclusions()
+        # The following dict uses product template attribute values as keys
+        # 0 means the value is acceptable, greater than 0 means it's rejected, it cannot be negative
+        # Bear in mind that several values can reject the same value and the latter can only be included in the
+        #  considered combination if no value rejects it.
+        # This dictionary counts how many times each value is rejected.
+        # Each time a value is included in the considered combination, the values it rejects are incremented
+        # When a value is discarded from the considered combination, the values it rejects are decremented
+        current_exclusions = defaultdict(int)
+        for exclusion in self._get_parent_attribute_exclusions(parent_combination):
+            current_exclusions[exclusion] += 1
+        included_values_ids = set()
+
+        # -1 means no value has been picked for the attribute in the current (partial) combination
+        values_idx = [-1] * len(product_template_attribute_values)
+        # determines which attribute line we're working on
+        attribute_idx = 0
+
+        def current_value_id():
+            return product_template_attribute_values[attribute_idx][values_idx[attribute_idx]].id
+
+        while True:
+            if values_idx[attribute_idx] >= 0:
+                for value_to_include_back in all_exclusions[current_value_id()]:
```

Comment:
Are you sure about `all_exclusions[current_value_id()]`?
I had the feeling the dict returned by `_get_own_attribute_exclusions` was indexed by record and not by record id.

It's actually very strange to ever need to work with ids directly instead of records.

Reply:
Ok nvm indeed it is indexed by ids, weird. I guess I would still prefer if we worked with records for as long as possible, and just convert to id explicitly here instead of inside helper.

Reply:
`self._get_own_attribute_exclusions()` returns `Dict[int, List[int]]` where `int`  are the ids of the `product_template_attribute_value`.
Working with the records would mean I would need to browse them and that would be a great waste of CPU time.

Reply:
The method returned ids to be callable by RPC, I don't think it's really a performance issue to use browse in python in this case, unless you have data that suggests otherwise.

Reply:
done